diff --git a/music_assistant/providers/snapcast/__init__.py b/music_assistant/providers/snapcast/__init__.py
index 2fd845ea..ae24c377 100644
--- a/music_assistant/providers/snapcast/__init__.py
+++ b/music_assistant/providers/snapcast/__init__.py
@@ -29,9 +29,13 @@ from music_assistant.providers.snapcast.constants import (
     CONF_SERVER_TRANSPORT_CODEC,
     CONF_STREAM_IDLE_THRESHOLD,
     CONF_USE_EXTERNAL_SERVER,
+    CONF_SNAPSERVER_STREAM_PORT_START,
+    CONF_SNAPSERVER_STREAM_PORT_END,
     DEFAULT_SNAPSERVER_IP,
     DEFAULT_SNAPSERVER_PORT,
     DEFAULT_SNAPSTREAM_IDLE_THRESHOLD,
+    DEFAULT_SNAPSERVER_STREAM_PORT_START,
+    DEFAULT_SNAPSERVER_STREAM_PORT_END,
 )
 from music_assistant.providers.snapcast.provider import SnapCastProvider
 
@@ -197,4 +201,24 @@ async def get_config_entries(
             required=True,
             category=CONF_CATEGORY_ADVANCED,
         ),
+        ConfigEntry(
+            key=CONF_SNAPSERVER_STREAM_PORT_START,
+            type=ConfigEntryType.INTEGER,
+            default_value=DEFAULT_SNAPSERVER_STREAM_PORT_START,
+            label="Music-assistant stream port range start (1 port used for each stream)",
+            required=False,
+            category=(
+                CONF_CATEGORY_ADVANCED if local_snapserver_present else CONF_CATEGORY_GENERIC
+            ),
+        ),
+        ConfigEntry(
+            key=CONF_SNAPSERVER_STREAM_PORT_END,
+            type=ConfigEntryType.INTEGER,
+            default_value=DEFAULT_SNAPSERVER_STREAM_PORT_END,
+            label="Music-assistant stream port range end (1 port used for each stream)",
+            required=False,
+            category=(
+                CONF_CATEGORY_ADVANCED if local_snapserver_present else CONF_CATEGORY_GENERIC
+            ),
+        ),
     )
diff --git a/music_assistant/providers/snapcast/constants.py b/music_assistant/providers/snapcast/constants.py
index eb5ecb1a..465d3845 100644
--- a/music_assistant/providers/snapcast/constants.py
+++ b/music_assistant/providers/snapcast/constants.py
@@ -17,6 +17,8 @@ CONF_SERVER_INITIAL_VOLUME = "snapcast_server_built_in_initial_volume"
 CONF_SERVER_TRANSPORT_CODEC = "snapcast_server_built_in_codec"
 CONF_SERVER_SEND_AUDIO_TO_MUTED = "snapcast_server_built_in_send_muted"
 CONF_STREAM_IDLE_THRESHOLD = "snapcast_stream_idle_threshold"
+CONF_SNAPSERVER_STREAM_PORT_START = "snapcast_stream_port_start"
+CONF_SNAPSERVER_STREAM_PORT_END = "snapcast_stream_port_end"
 
 
 CONF_CATEGORY_GENERIC = "generic"
@@ -35,6 +37,8 @@ CONF_ENTRY_SAMPLE_RATES_SNAPCAST = create_sample_rates_config_entry(
 DEFAULT_SNAPSERVER_IP = "127.0.0.1"
 DEFAULT_SNAPSERVER_PORT = 1705
 DEFAULT_SNAPSTREAM_IDLE_THRESHOLD = 60000
+DEFAULT_SNAPSERVER_STREAM_PORT_START = 4953
+DEFAULT_SNAPSERVER_STREAM_PORT_END = 5153
 
 # Socket path template for control script communication
 # The {queue_id} placeholder will be replaced with the actual queue ID
diff --git a/music_assistant/providers/snapcast/player.py b/music_assistant/providers/snapcast/player.py
index 011250e6..8c52980b 100644
--- a/music_assistant/providers/snapcast/player.py
+++ b/music_assistant/providers/snapcast/player.py
@@ -18,6 +18,8 @@ from music_assistant.constants import (
     ATTR_ANNOUNCEMENT_IN_PROGRESS,
     CONF_ENTRY_FLOW_MODE_ENFORCED,
     CONF_ENTRY_OUTPUT_CODEC_HIDDEN,
+    CONF_SNAPSERVER_STREAM_PORT_START,
+    CONF_SNAPSERVER_STREAM_PORT_END,
 )
 from music_assistant.helpers.audio import get_player_filter_params
 from music_assistant.helpers.compare import create_safe_string
@@ -342,6 +344,15 @@ class SnapCastPlayer(Player):
             stream_name += MASS_ANNOUNCEMENT_POSTFIX
         return stream_name
 
+    def _get_port(self) -> int:
+        start = self.provider.config.get_value(CONF_SNAPSERVER_STREAM_PORT_START)
+        end = self.provider.config.get_value(CONF_SNAPSERVER_STREAM_PORT_END)
+        if start == end:
+            return start
+        if start > end:
+            start, end = end, start # user specified range counter-intuitively
+        return random.randint(start, end)
+
     async def _get_or_create_stream(self, stream_name: str, queue_id: str | None) -> Snapstream:
         """Create new stream on snapcast server (or return existing one)."""
         # prefer to reuse existing stream if possible
@@ -369,7 +380,8 @@ class SnapCastPlayer(Player):
         while attempts:
             attempts -= 1
             # pick a random port
-            port = random.randint(4953, 4953 + 200)
+            port = self._get_port()
+            self.logger.debug("Requesting port %i (configured to use between %i and %i)", port, self.provider.config.get_value(CONF_SNAPSERVER_STREAM_PORT_START), self.provider.config.get_value(CONF_SNAPSERVER_STREAM_PORT_END))
             result = await self.provider._snapserver.stream_add_stream(
                 # NOTE: setting the sampleformat to something else
                 # (like 24 bits bit depth) does not seem to work at all!
