diff --git a/.gitignore b/.gitignore
index 673eb1d9..efef05e7 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,5 @@
+/.config
+/shell.nix
 
 .DS_Store
 *.db
diff --git a/music_assistant/providers/refreezer/__init__.py b/music_assistant/providers/refreezer/__init__.py
new file mode 100644
index 00000000..055e1e16
--- /dev/null
+++ b/music_assistant/providers/refreezer/__init__.py
@@ -0,0 +1,1050 @@
+from __future__ import annotations
+
+import asyncio
+
+from collections.abc import AsyncGenerator, Sequence
+from typing import TYPE_CHECKING
+
+from hashlib import md5
+import urllib.parse
+import uuid
+from Crypto.Cipher import Blowfish
+
+from aiohttp import ClientSession, ClientTimeout, ClientResponse
+
+from music_assistant_models.config_entries import ConfigEntry, ConfigValueType, ConfigValueOption, ProviderConfig
+from music_assistant_models.enums import (
+    AlbumType,
+    ConfigEntryType,
+    ContentType,
+    ExternalID,
+    ImageType,
+    MediaType,
+    ProviderFeature,
+    StreamType,
+)
+from music_assistant_models.errors import UnplayableMediaError
+from music_assistant_models.media_items import (
+    Album,
+    Artist,
+    AudioFormat,
+    ItemMapping,
+    MediaItemImage,
+    MediaItemMetadata,
+    MediaItemType,
+    Playlist,
+    ProviderMapping,
+    RecommendationFolder,
+    SearchResults,
+    Track,
+)
+from music_assistant_models.provider import ProviderManifest
+from music_assistant_models.streamdetails import StreamDetails
+
+from music_assistant import MusicAssistant
+from music_assistant.helpers.app_vars import app_var
+from music_assistant.helpers.auth import AuthenticationHelper
+from music_assistant.helpers.datetime import utc_timestamp
+from music_assistant.models import ProviderInstanceType
+from music_assistant.models.music_provider import MusicProvider
+
+import random
+import json
+import logging
+from typing import Optional, Dict, Any, List
+
+
+
+SUPPORTED_FEATURES = {
+    ProviderFeature.LIBRARY_ARTISTS,
+    ProviderFeature.LIBRARY_ALBUMS,
+    ProviderFeature.LIBRARY_TRACKS,
+    ProviderFeature.LIBRARY_PLAYLISTS,
+    ProviderFeature.LIBRARY_ALBUMS_EDIT,
+    ProviderFeature.LIBRARY_TRACKS_EDIT,
+    ProviderFeature.LIBRARY_ARTISTS_EDIT,
+    ProviderFeature.LIBRARY_PLAYLISTS_EDIT,
+    ProviderFeature.ALBUM_METADATA,
+    ProviderFeature.TRACK_METADATA,
+    ProviderFeature.ARTIST_METADATA,
+    ProviderFeature.ARTIST_ALBUMS,
+    ProviderFeature.ARTIST_TOPTRACKS,
+    ProviderFeature.BROWSE,
+    ProviderFeature.SEARCH,
+    ProviderFeature.PLAYLIST_TRACKS_EDIT,
+    ProviderFeature.PLAYLIST_CREATE,
+    # ProviderFeature.RECOMMENDATIONS,
+    # ProviderFeature.SIMILAR_TRACKS,
+}
+
+CONF_REFREEZER_MAIL = "refreezer_mail"
+CONF_REFREEZER_PASS = "refreezer_pass"
+CONF_REFREEZER_ARL = "refreezer_arl"
+CONF_REFREEZER_QUALITY = "refreezer_quality"
+
+CONF_ACTION_AUTH = "arl_token"
+
+
+DEFAULT_HEADERS = {
+    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36',
+    'X-User-IP': '1.1.1.1',
+    'x-deezer-client-ip': '1.1.1.1',
+    'Accept': '*/*'
+}
+
+class CookieManager:
+    '''handles obtaining session cookies from deezer'''
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self.cookie_header = {};
+        self.cookies = {};
+
+    def update_cookie(self, r: ClientResponse):
+        '''updates cookie info'''
+        if not any(r.headers['set-cookie']):
+            return
+        setcookies = r.headers.getall('set-cookie')
+        for setc in setcookies:
+            cookies = setc.split(';')
+            for c in cookies:
+                self._set_cookie(c)
+        self.cookie_header['cookie'] = self._generate_cookie_header();
+
+    def _set_cookie(self, cookie: str):
+        if (not cookie) or (not '=' in cookie):
+            return
+        key, value = [s.strip() for s in cookie.split('=', 1)]
+        if key.lower() in ['path', 'expires', 'domain', 'samesite']:
+            return # not interesting
+        # if we get here, this is an interesting cookie - keep it
+        self.cookies[key] = value
+
+    def _generate_cookie_header(self):
+        return ';'.join((f'{k}={v}'for k, v in self.cookies.items()))
+
+
+async def get_arl_by_email_and_password(email: str, password: str, http_session: ClientSession) -> str:
+    '''Automatically returns ARL token by using the deezer API'''
+    cookie_manager = CookieManager()
+    # remove any existing cookies from http_session
+    http_session.cookie_jar.clear()
+
+    # get initial cookies (sid) from empty getUser call
+    response = await http_session.get('https://www.deezer.com/ajax/gw-light.php?method=deezer.getUserData&input=3&api_version=1.0&api_token=null')
+    cookie_manager.update_cookie(response)
+    http_session.cookie_jar.clear()
+
+    # Try to get AccessToken by login with email & password, which sets authentication cookies
+    token = await _get_accesstoken(email, password, cookie_manager, http_session)
+    if (token == None):
+        return ''
+
+    # get arl
+    request_headers = DEFAULT_HEADERS | cookie_manager.cookie_header
+    response = await http_session.get('https://www.deezer.com/ajax/gw-light.php?method=user.getArl&input=3&api_version=1.0&api_token=null', headers=request_headers)
+
+    data = await response.json()
+    return data['results']
+
+async def _get_accesstoken(email: str, password: str, cookie_manager: CookieManager, http_session: ClientSession):
+    # music-assistant
+    # DEEZER_APP_ID = '600344'
+    # DEEZER_APP_S = '1e79e9ecb000010147b7e31c21794b74'
+    # Spleeter, Deezloader creds (maybe)
+    # DEEZER_APP_ID = '243762'
+    # DEEZER_APP_S = 'a5cf06353c1d7083e2fca5a6885a1a74'
+    # Deezer API (same as streamrip & Orpheus)
+    DEEZER_APP_ID = '447462'
+    DEEZER_APP_S = 'a83bf7f38ad2f137e444727cfc3775cf'
+    # LastFM API original freezer
+    # LASTFM_API_ID = 'b6ab5ae967bcd8b10b23f68f42493829'
+    # LASTFM_API_S = '861b0dff9a8a574bec747f9dab8b82bf'
+
+    request_headers = DEFAULT_HEADERS | cookie_manager.cookie_header
+    hashed_pass = md5(password.encode('utf-8')).hexdigest()
+    hashed_params = md5(f'{DEEZER_APP_ID}{email}{hashed_pass}{DEEZER_APP_S}'.encode('utf-8')).hexdigest()
+
+    params = {'app_id': DEEZER_APP_ID, 'login': email, 'password': hashed_pass, 'hash': hashed_params}
+    url = f'https://connect.deezer.com/oauth/user_auth.php?{urllib.parse.urlencode(params)}'
+
+    response = await http_session.get(url, headers=request_headers)
+    cookie_manager.update_cookie(response)
+    json = await response.json()
+    if 'access_token' in json:
+        return json['access_token']
+    elif 'error' in json and any(json['error']):
+        raise ValueError(f'got an error of type \'{json["error"]["type"]}\', message: \'{json["error"]["message"]}\'')
+    return None
+
+
+async def setup(mass: MusicAssistant, manifest: ProviderManifest, config: ProviderConfig) -> ProviderInstanceType:
+    """Initialize provider(instance) with given configuration."""
+    return ReFreezerProvider(mass, manifest, config)
+
+
+async def get_config_entries(
+    mass: MusicAssistant,
+    instance_id: str | None = None,
+    action: str | None = None,
+    values: dict[str, ConfigValueType] | None = None,
+) -> tuple[ConfigEntry, ...]:
+    """
+    Return Config entries to setup this provider.
+
+    instance_id: id of an existing provider instance (None if new instance setup).
+    action: [optional] action key called from config entries UI.
+    values: the (intermediate) raw values for config entries sent with the action.
+    """
+    if action == CONF_ACTION_AUTH:
+        values[CONF_REFREEZER_ARL] = await get_arl_by_email_and_password(values[CONF_REFREEZER_MAIL], values[CONF_REFREEZER_PASS], mass.http_session)
+
+    return (
+        ConfigEntry(
+            key=CONF_REFREEZER_MAIL,
+            type=ConfigEntryType.STRING,
+            label="email",
+            required=True,
+            description="Deezer email",
+            value=values.get(CONF_REFREEZER_MAIL) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_REFREEZER_PASS,
+            type=ConfigEntryType.SECURE_STRING,
+            label="password",
+            required=True,
+            description="Deezer password",
+            value=values.get(CONF_REFREEZER_PASS) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_REFREEZER_QUALITY,
+            type=ConfigEntryType.STRING,
+            options=[
+                ConfigValueOption('FLAC (premium-only)', 'FLAC'),
+                ConfigValueOption('MP3 320kbit (premium-only)', 'MP3_320'),
+                ConfigValueOption('MP3 128kbit', 'MP3_128'),
+            ],
+            label="Stream quality",
+            required=True,
+            description="Stream quality (FLAC only works for premium deezer accounts)",
+            value=values.get(CONF_REFREEZER_QUALITY) if values else None,
+            default_value = "MP3_128",
+        ),
+        ConfigEntry(
+            key=CONF_REFREEZER_ARL,
+            type=ConfigEntryType.SECURE_STRING,
+            label="Get ARL token",
+            required=True,
+            action=CONF_ACTION_AUTH,
+            description="Deezer ARL token",
+            value=values.get(CONF_REFREEZER_ARL) if values else None,
+        ),
+    )
+
+class DeezerAPI:
+    class Settings:
+        defaults = {
+            'deezer_language': 'en',
+            'deezer_country': 'US'
+        }
+        def __init__(self, **kwargs):
+            '''populate settings using passed kwargs, default value otherwise'''
+            pick = lambda name: kwargs[name] if name in kwargs else DeezerAPI.Settings.defaults[name]
+            self.deezer_language: str = pick('deezer_language')
+            self.deezer_country: str = pick('deezer_country')
+
+    class ImageDetails:
+        def __init__(self, full_url: str, thumb_url: str, type: str, image_hash: str):
+            self.full_url = full_url
+            self.thumb_url = thumb_url
+            self.type = type
+            self.image_hash = image_hash
+
+        def from_private_string(image_hash: str, type: str = 'cover'):
+            return DeezerAPI.ImageDetails(full_url=DeezerAPI.ImageDetails.custom_url(image_hash, type, 1000, 1000, 80), thumb_url=DeezerAPI.ImageDetails.custom_url(image_hash, type, 140, 140, 80), type=type, image_hash=image_hash)
+
+        def from_private_json(data: Dict):
+            return DeezerAPI.ImageDetails.from_private_string(data['MD5'] if 'MD5' in data and data['MD5'] != None else data['md5'], data['TYPE'] if 'TYPE' in data and data['TYPE'] != None else data['type'])
+
+        def custom_url(image_hash: str, type: str, height: int, width: int, quality: int) -> str:
+            return f'https://e-cdns-images.dzcdn.net/images/{type}/{image_hash}/{height}x{width}-000000-{quality}-0-0.jpg'
+
+        def url(self, height: int, width: int, quality: int) -> str:
+            return DeezerAPI.ImageDetail.custom_url(self.image_hash, self.type, height, width, quality)
+
+
+    user_agent = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36'
+
+    def __init__(self, session: ClientSession, logger: logging.Logger, arl: Optional[str] = None, settings: Optional[Settings] = None):
+        self.session: ClientSession = session
+        self.logger: logging.Logger = logger
+        self.arl = arl
+        self.token: Optional[str] = None
+        self.license_token: Optional[str] = None
+        self.user_id: Optional[str] = None
+        self.user_name: Optional[str] = None
+        self.favorites_playlist_id: Optional[str] = None
+        self.sid: Optional[str] = None
+        self._authorizing: Optional[asyncio.Future] = None
+        self.settings: Settings = settings or DeezerAPI.Settings()
+
+    def get_headers(self) -> Dict[str, str]:
+        lang = f'{self.settings.deezer_language}-{self.settings.deezer_country}'
+        cookie = f'arl={self.arl}' if self.sid == None else f'arl={self.arl}; sid={self.sid}'
+        return {
+            'User-Agent': self.user_agent,
+            'Content-Language': lang,
+            'Content-Type': 'text/plain;charset=UTF-8',
+            'Accept': '*/*',
+            'Accept-Charset': 'utf-8,ISO-8859-1;q=0.7,*;q=0.3',
+            'Accept-Language': f'{lang},{self.settings.deezer_language};q=0.9,en-US;q=0.8,en;q=0.7',
+            'Connection': 'keep-alive',
+            'Cookie': cookie,
+        }
+
+    async def call_gw_api(self, method: str, params: Optional[Dict] = None, gateway_input: Optional[str] = None) -> Dict[str, Any]:
+        url = 'https://www.deezer.com/ajax/gw-light.php'
+        query = {
+            'api_version': '1.0',
+            'api_token': self.token or '', # must provide a string, and null params should mean that a key is not set
+            'input': '3',
+            'method': method,
+            'cid': str(random.randint(0, 1_000_000_000)),
+        }
+        if gateway_input:
+            query['gateway_input'] = gateway_input
+
+        headers = self.get_headers()
+        async with self.session.post(url, headers=headers, params=query, data=json.dumps(params or {})) as resp:
+            body = await resp.json()
+
+            # Grab SID cookie
+            if method == 'deezer.getUserData':
+                set_cookie = resp.headers.get('Set-Cookie', '')
+                for cookie in set_cookie.split(';'):
+                    if cookie.strip().startswith('sid='):
+                        self.sid = cookie.strip().split('=', 1)[1]
+
+            # In case of error 'Invalid CSRF token' retrieve new one and retry the same call
+            # Except for 'deezer.getUserData' method, which would cause infinite loop
+            if ( 'error' in body and 'VALID_TOKEN_REQUIRED' in body['error'] and method != 'deezer.getUserData' and await self.raw_authorize() ):
+                return await self.call_gw_api(method, params=params, gateway_input=gateway_input)
+            return body
+
+    async def call_public_api(self, path: str) -> Dict[str, Any]:
+        async with self.session.get(f'https://api.deezer.com/{path}') as resp:
+            return await resp.json()
+
+    async def get_json_web_token(self) -> str:
+        url = 'https://auth.deezer.com/login/arl'
+        params = {'jo': 'p', 'rto': 'c', 'i': 'c'}
+        headers = self.get_headers()
+        async with self.session.post(url, headers=headers, params=params) as resp:
+            data = await resp.json()
+            return data.get('jwt', '')
+
+    async def call_pipe_api(self, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
+        jwt_token = await self.get_json_web_token()
+        headers = self.get_headers()
+        headers['Authorization'] = f'Bearer {jwt_token}'
+        headers['Content-Type'] = 'application/json'
+        url = 'https://pipe.deezer.com/api/'
+        async with self.session.post(url, headers=headers, data=json.dumps(params or {})) as resp:
+            return await resp.json()
+
+    async def authorize(self) -> bool:
+        if self._authorizing is None:
+            self._authorizing = asyncio.ensure_future(self.raw_authorize())
+            result = await self._authorizing
+            self._authorizing = None
+            return result
+        return await self._authorizing
+
+    async def raw_authorize(self) -> bool:
+        try:
+            data = await self.call_gw_api("deezer.getUserData")
+            user = data.get("results", {}).get("USER", {})
+            if user.get("USER_ID") == 0:
+                self.logger.error(f"I don't even know how I got here: user raw data was: {data}")
+                return False
+
+            self.token = data["results"].get("checkForm")
+            self.user_id = str(user.get("USER_ID", ""))
+            self.user_name = user.get("BLOG_NAME")
+            self.favorites_playlist_id = user.get("LOVEDTRACKS_ID")
+            self.license_token = user.get("OPTIONS", {}).get("license_token")
+            return True
+        except Exception as e:
+            self.logger.error("Login Error (D): %s", e)
+            return False
+
+# seb TODO:
+# parseLink
+# checkAvailability
+
+
+    async def search(self, query: str, start: int = 0, nb: int = 128) -> tuple[List[Dict[str, Any]], List[Dict[str, Any]], List[Dict[str, Any]], List[Dict[str, Any]]]:
+        '''
+        search deezer on the given query. Searches through all categories (e.g. artists, tracknames etc) at once.
+        Returns 4 lists with json data: tracks, albums, artists, playlists (in this order).
+        '''
+        data = await self.call_gw_api('deezer.pageSearch', params={'nb': nb, 'query': query, 'start': start})
+        if not 'results' in data:
+            return [], [], [], []
+        tracks = data['results'].get('TRACK', {}).get('data', [])
+        albums = data['results'].get('ALBUM', {}).get('data', [])
+        artists= data['results'].get('ARTIST', {}).get('data', [])
+        playlists = data['results'].get('PLAYLIST', {}).get('data', [])
+        return tracks, albums, artists, playlists
+
+    async def track(self, id: str) -> Dict[str, Any]:
+        data = await self.call_gw_api('song.getListData', params={'sng_ids': [id]})
+        return data['results']['data'][0]
+
+    async def album(self, id: str) -> Dict[str, Any]:
+        data = await self.call_gw_api('deezer.pageAlbum', params={'alb_id': id, 'header': True, 'lang': self.settings.deezer_language})
+        return data['results']
+
+    async def artist(self, id: str) -> Dict[str, Any]:
+        data = await self.call_gw_api('deezer.pageArtist', params={'art_id': id, 'lang': self.settings.deezer_language})
+        return data['results']
+
+    async def playlist_track_page(self, id: str, start: int, nb: int = 50) -> List[Dict[str, Any]]:
+        '''get playlist songs, from `start` up to `start+nb`. Returns a list of json data that can be directly converted to tracks.'''
+        data = await self.call_gw_api('deezer.pagePlaylist', params={'playlist_id': id, 'lang': self.settings.deezer_language, 'nb': nb, 'tags': True, 'start': start})
+        return data['results']['SONGS']['data']
+
+    async def playlist(self, id: str, nb: int = 100) -> Dict[str, Any]:
+        '''get playlist by id, with up to `nb` tracks'''
+        data = await self.call_gw_api('deezer.pagePlaylist', params={'playlist_id': id, 'lang': self.settings.deezer_language, 'nb': nb, 'tags': True, 'start': 0})
+        return data['results']
+
+    async def full_playlist(self, id: str) -> Dict[str, Any]:
+        '''get playlist by id, with all tracks'''
+        return await self.playlist(id, nb=100000)
+
+    async def add_favorite_track(self, id: str):
+        await self.call_gw_api('favorite_song.add', params={'SNG_ID': id})
+
+    async def add_favorite_album(self, id: str):
+        '''mark an album as favorite'''
+        await self.call_gw_api('album.addFavorite', params={'ALBUM_ID': id})
+
+    async def add_favorite_artist(self, id: str):
+        '''mark an artist as favorite.'''
+        await self.call_gw_api('artist.addFavorite', params={'ARTIST_ID': id})
+
+    async def remove_artist(self, id: str):
+        '''remove an artist from favorites.'''
+        await self.call_gw_api('artist.removeFavorite', params={'ARTIST_ID': id})
+
+    async def dislike_track(self, id: str):
+        await self.call_gw_api('favorite_dislike.add', params={'ID': id, 'TYPE': 'song'})
+
+    async def add_to_playlist(self, track_id: str, playlist_id: str, offset: int=-1):
+        ''' add a song to a playlist'''
+        await self.call_gw_api('playlist.addSongs', params={'offset': offset, 'playlist_id': playlist_id, 'songs': [[track_id, 0]]})
+
+    async def remove_from_playlist(self, track_id: str, playlist_id: str):
+        ''' remove a song from a playlist'''
+        await self.call_gw_api('playlist.deleteSongs', params={'playlist_id': playlist_id, 'songs': [[track_id, 0]]})
+
+    async def create_playlist(self, title: str, description: str='', status: int=1, track_ids: List[str]=[]) -> str:
+        '''Create a playlist with a list of songs'''
+        songs = [[int(tid), idx] for idx, tid in enumerate(track_ids)]
+        data = await self.call_gw_api('playlist.create', params={'title': title, 'description': description, 'songs': songs, 'status': status})
+        return str(data['results'])
+
+    async def remove_playlist(self, id: str, owned_by_us: Optional[bool] = None):
+        '''
+        Delete a playlist (if user owns it) or remove it from favorites (if user does not own it).
+        Leave `owned_by_us` empty to let the API figure it out, at the cost of an extra API call.
+        '''
+        if owned_by_us == None:
+            playlist_raw = await self.playlist(id, nb=0)
+            owned_by_us = playlist_raw['results']['DATA']['PARENT_USER_ID'].strip() == self.user_id if playlist_raw['results']['DATA']['PARENT_USER_ID'] != None else False # no parent id --> not made by us
+
+        if owned_by_us:
+            await self.call_gw_api('playlist.delete', params={'playlist_id': id}) # full deletion
+        else:
+            await self.call_gw_api('playlist.deleteFavorite', params={'playlist_id': id}) # we don't own it, remove from library/favorites
+
+    async def update_playlist(self, id: str, title: str, description: str, status: int = 1):
+        await self.call_gw_api('playlist.update', params={'description': description, 'title': title, 'playlist_id': int(id), 'status': status, 'songs': []})
+
+    async def get_playlists(self, nb: int = 100):
+        data = await self.call_gw_api('deezer.pageProfile', params={'nb': nb, 'tab': 'playlists', 'user_id': self.user_id});
+        return data['results']['TAB']['playlists']['data']
+
+    async def get_favorite_track_ids(self):
+        data = await self.call_gw_api('user.getAllFeedbacks', params={'checksums': None})
+        if data['results'] == None or data['results']['FAVORITES'] == None or data['results']['FAVORITES']['SONGS'] == None:
+            return None # we get here when e.g. user has no favorites yet
+        return [x['SNG_ID'] for x in data['results']['FAVORITES']['SONGS']['data']]
+
+    async def get_albums(self, nb: int=50):
+        data = await self.call_gw_api('deezer.pageProfile', params={'nb': nb, 'tab': 'albums', 'user_id': self.user_id})
+        return data['results']['TAB']['albums']['data']
+
+    async def remove_album(self, id: str) -> None:
+        '''remove an album from favorites.'''
+        await self.call_gw_api('album.removeFavorite', params={'ALBUM_ID': id})
+
+    async def remove_favorite_song(self, id: str) -> None:
+        '''remove a track from favorites.'''
+        await self.call_gw_api('song.removeFavorite', params={'SNG_ID': id})
+
+    async def get_artists(self, nb: int=40):
+        '''returns favorite/library artists, up to `nb` items. Can be directly converted to artists.'''
+        data = await self.call_gw_api('deezer.pageProfile', params={'nb': nb, 'tab': 'artists', 'user_id': self.user_id})
+        return data['results']['TAB']['artists']['data']
+
+
+# lyrics
+# lyricsLegacy
+# lyricsFull
+# smartTrackList
+# flow
+# homePage
+
+    async def log_listen(self, id: str) -> None:
+        '''log that a track has been listened to'''
+        await self.call_gw_api('log.listen', params={'SNG_ID': id})
+
+# getChannel
+
+    async def add_playlist(self, id: str) -> None:
+        '''Add a playlist to favorites from some parent playlist'''
+        await self.call_gw_api('playlist.addFavorite', params={'parent_playlist_id': id})
+
+# discographyPage
+# searchSuggestions
+# smartRadio
+# libraryShuffle
+# playMix
+# allShowEpisodes
+
+    async def get_track_url(self, track_id: str, track_token: str, md5origin: str, media_version: str, quality: str, refresh_attempt: int=0) -> tuple[Optional[str], bool]:
+        '''returns a (url, encrypted) pair for streaming a track. url is a direct media link.'''
+        if self.license_token is None:
+            self.logger.error(f'License token is None, how can that be? I need it here')
+            return None, True
+
+        if quality not in ['FLAC', 'MP3_320', 'MP3_128']:
+            self.loger.error(f'Found unknown quality type "{quality}"')
+            return None, True
+        format = quality
+
+        payload = {
+            "license_token": self.license_token,
+            "media": [ { "type": "FULL", "formats": [{"cipher": "BF_CBC_STRIPE", "format": format}], } ],
+            "track_tokens": [track_token],
+        }
+
+        headers = self.get_headers()
+        headers["Content-Type"] = "application/json"
+
+        try:
+            async with self.session.post('https://media.deezer.com/v1/get_url', headers=headers, data=json.dumps(payload)) as resp:
+                result = await resp.json()
+
+            if 'errors' in result:
+                raise ValueError(f'Encountered the following errors: {result['errors']}')
+
+            if 'data' in result:
+                for data in result['data']:
+                    if 'errors' in data:
+                        for error in data['errors']:
+                            if error.get('code') == 2001 and refresh_attempt < 1:
+                                # track token expired, so refresh track metadata and retry once
+                                track_data = await self.track(track_id)
+                                private_track = track_data['results']['data'][0]
+                                return await self.get_track_url(
+                                    private_track['SNG_ID'],
+                                    private_track['TRACK_TOKEN'],
+                                    private_track['MD5_ORIGIN'],
+                                    private_track['MEDIA_VERSION'],
+                                    quality,
+                                    refresh_attempt + 1,
+                                )
+                        self.logger.warning('Failed in getting streaming URL: %s', data.get('errors'))
+                    if 'media' in data and data['media']:
+                        from pprint import pformat
+                        self.logger.info(f'got data url etc: {data["media"][0]}')
+                        url = data['media'][0]['sources'][0]['url']
+                        return url, True
+        except Exception as e:
+            self.logger.warning("Error getting streaming URL: %s", e)
+        return None, True
+
+
+class ReFreezerProvider(MusicProvider):
+    """ReFreezer provider support."""
+
+    api: DeezerAPI
+
+    async def handle_async_init(self, settings: Optional[DeezerAPI.Settings] = None) -> None:
+        self.api = DeezerAPI(session=self.mass.http_session, logger=self.logger, arl=self.config.get_value(CONF_REFREEZER_ARL), settings=settings)
+        if not await self.api.authorize():
+            self.logger.error('Could not authorize with deezer')
+
+
+    @property
+    def supported_features(self) -> set[ProviderFeature]:
+        """Return the features supported by this Provider."""
+        return SUPPORTED_FEATURES
+
+    async def search(self, search_query: str, media_types: list[MediaType], limit: int = 30) -> SearchResults:
+        """Perform search on musicprovider.
+
+        :param search_query: Search query.
+        :param media_types: A list of media_types to include.
+        :param limit: Number of items to return in the search (per type).
+        """
+        tracks, albums, artists, playlists = await self.api.search(search_query, nb=limit)
+
+        results = SearchResults()
+        if MediaType.TRACK in media_types:
+            results.tracks = self._parse_tracks(tracks)
+            # seb TODO: down here:
+            # Aug 30 19:41:22 helium mass[73093]:   File "/nix/store/dbcn999rhl0f4f3hnl95hkrbwig08hlp-music-assistant-2.5.8/lib/python3.13/site-packages/music_assistant/providers/refreezer/__init__.py", line 538, in search
+            # Aug 30 19:41:22 helium mass[73093]:     results.tracks += self._parse_tracks(tracks)
+            # Aug 30 19:41:22 helium mass[73093]:                       ~~~~~~~~~~~~~~~~~~^^^^^^^^
+            # Aug 30 19:41:22 helium mass[73093]:   File "/nix/store/dbcn999rhl0f4f3hnl95hkrbwig08hlp-music-assistant-2.5.8/lib/python3.13/site-packages/music_assistant/providers/refreezer/__init__.py", line 754, in _parse_tr>
+            # Aug 30 19:41:22 helium mass[73093]:     return [self._parse_track(elem, album=album) for elem in data['data']] if any(data['data']) else []
+            # Aug 30 19:41:22 helium mass[73093]:                                                                                   ~~~~^^^^^^^^
+            # Aug 30 19:41:22 helium mass[73093]: TypeError: list indices must be integers or slices, not str
+            # that was in the _parse_tracks function
+        if MediaType.ALBUM in media_types:
+            results.albums = self._parse_albums(albums)
+        if MediaType.ARTIST in media_types:
+            results.artists = self._parse_artists(artists)
+        if MediaType.PLAYLIST in media_types:
+            results.playlists = self._parse_playlists(playlists)
+        return results
+
+    async def get_library_artists(self) -> AsyncGenerator[Artist, None]:
+        """Retrieve library artists from the provider."""
+        # seb NOTE: only gets first 40 artists
+        # return (self._parse_artist(x) for x in await self.api.get_artists(nb=40))
+        self.logger.error(f'Got a artist lib request')
+        data = await self.api.get_artists(nb=40)
+        for x in data:
+            yield self._parse_artist(x)
+
+    async def get_library_albums(self) -> AsyncGenerator[Album, None]:
+        """Retrieve library albums from the provider."""
+        # seb NOTE: only gets first 50 albums
+        self.logger.error(f'Got a album lib request')
+        data = await self.api.get_albums(nb=50)
+        # yield from map(self._parse_album, await self.api.get_albums(nb=50))
+        for x in data:
+            yield self._parse_album(x)
+        # async for x in await self.api.get_albums(nb=40):
+        #     yield self._parse_album(x)
+
+    async def get_library_tracks(self) -> AsyncGenerator[Track, None]:
+        """Retrieve library tracks from the provider."""
+        # return (await self.get_track(id) for id in await self.api.get_favorite_track_ids()) # must return generator
+        self.logger.error(f'Got a track lib request - seb TODO: This is too slow (use nice async generator chain as in spotify)')
+        data = [await self.api.track(x) for x in await self.api.get_favorite_track_ids()]
+        for x in data:
+            yield self._parse_track(x)
+
+    async def get_library_playlists(self) -> AsyncGenerator[Playlist, None]:
+        """Retrieve library/subscribed playlists from the provider."""
+        self.logger.error(f'Got a playlist lib request')
+        # seb NOTE: only gets first 100 playlists
+        # seb NOTE: problem looks like:
+        # Aug 30 08:14:12 helium mass[15490]: TypeError: 'async for' requires an object with __aiter__ method, got coroutine
+        # Aug 30 08:14:12 helium mass[15490]: 2025-08-30 08:14:12.428 WARNING (MainThread) [music_assistant.music] Sync task for ReFreezer completed with errors
+        # Aug 30 08:14:12 helium mass[15490]: Traceback (most recent call last):
+        # Aug 30 08:14:12 helium mass[15490]:   File "/nix/store/2l09xhxpdys9ija9gy6f4ny997w5jjhl-music-assistant-2.5.8/lib/python3.13/site-packages/music_assistant/controllers/music.py", line 1235, in run_sync
+        # Aug 30 08:14:12 helium mass[15490]:     await provider.sync_library(media_type)
+        # Aug 30 08:14:12 helium mass[15490]:   File "/nix/store/2l09xhxpdys9ija9gy6f4ny997w5jjhl-music-assistant-2.5.8/lib/python3.13/site-packages/music_assistant/models/music_provider.py", line 632, in sync_library
+        # Aug 30 08:14:12 helium mass[15490]:     async for prov_item in self._get_library_gen(media_type):
+        # Aug 30 08:14:12 helium mass[15490]:                            ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
+        # Aug 30 08:14:12 helium mass[15490]: TypeError: 'async for' requires an object with __aiter__ method, got coroutine
+        # return (self._parse_playlist(x) for x in await self.api.get_playlists(nb=100)) # must return generator
+        for x in await self.api.get_playlists(nb=100):
+            yield self._parse_playlist(x)
+    
+    async def get_artist(self, prov_artist_id: str) -> Artist:
+        """Get full artist details by id."""
+        data = await self.api.artist(id=prov_artist_id)
+        return self._parse_artist(data['results']['DATA'])
+
+    async def get_artist_albums(self, prov_artist_id: str) -> list[Album]:
+        """Get a list of all albums for the given artist."""
+        data = await self.api.artist(id=prov_artist_id)
+        return self._parse_albums(data['results']['ALBUMS'])
+
+    async def get_artist_toptracks(self, prov_artist_id: str) -> list[Track]:
+        """Get a list of most popular tracks for the given artist."""
+        data = await self.api.artist(id=prov_artist_id)
+        return self._parse_tracks(data['results']['TOP'])
+
+    async def get_album(self, prov_album_id: str) -> Album:
+        """Get full album details by id."""
+        data = await self.api.album(id=prov_album_id)
+        # NOTE: we also have data['SONGS'] here, with all songs of the album. (data['SONGS']['data'] contains a list of songs, there also is a 'count')
+        # However, ma asks for this later in its `get_album_tracks` function.
+        # seb TODO: It is probably best to make a 'cache' object to maintain state of recently-requested album ids with their songs stored.
+        return self._parse_album(data['DATA'])
+
+    async def get_album_tracks( self, prov_album_id: str) -> list[Track]:
+        """Get album tracks for given album id."""
+        data = await self.api.album(id=prov_album_id)
+        return self._parse_tracks(data['SONGS']['data'])
+
+    async def get_track(self, prov_track_id: str) -> Track:
+        """Get full track details by id."""
+        data = await self.api.track(id=prov_track_id)
+        return self._parse_track(data)
+
+    async def get_playlist(self, prov_playlist_id: str) -> Playlist:
+        """Get full playlist details by id."""
+        data = await self.api.playlist(id=prov_playlist_id, nb=10) # seb TODO: does this mean it only loads the first 10 playlists and I can never get the rest?
+        # seb TODO: before used this expression (but did not use songs in parse function itself): return self._parse_playlist(data=data['results']['DATA'], songs=data['results']['SONGS'])
+        return self._parse_playlist(data=data['DATA'])
+
+    async def get_playlist_tracks(self, prov_playlist_id: str, page: int=0) -> list[Track]:
+        """Get all playlist tracks for given playlist id."""
+        page_size = 50
+        offset = page * page_size
+        data = await self.api.playlist_track_page(id=prov_playlist_id, start=offset, nb=page_size)
+        return self._parse_tracks(data)
+
+    async def library_add(self, item: MediaItemType) -> bool:
+        """Add item to provider's library. Return true on success."""
+        try:
+            if item.media_type == MediaType.ARTIST:
+                await self.api.add_favorite_artist(id=item.item_id)
+            elif item.media_type == MediaType.ALBUM:
+                await self.api.add_favorite_album(id=item.item_id)
+            elif item.media_type == MediaType.TRACK:
+                await self.api.add_favorite_track(id=item.item_id)
+            elif item.media_type == MediaType.PLAYLIST:
+                await self.api.add_playlist(id=item.item_id)
+            else:
+                raise ValueError("Unknown item media type: %s", str(item.media_type))
+        except Exception as e:
+            self.logger.error(e)
+            return False
+        return True
+
+    async def library_remove(self, prov_item_id: str, media_type: MediaType) -> bool:
+        """Remove item from provider's library. Return true on success."""
+        try:
+            if media_type == MediaType.ARTIST:
+                await self.api.remove_artist(id=prov_item_id)
+            elif media_type == MediaType.ALBUM:
+                await self.api.remove_album(id=prov_item_id)
+            elif media_type == MediaType.TRACK:
+                await self.api.remove_favorite_song(id=prov_item_id)
+            elif media_type == MediaType.PLAYLIST:
+                await self.api.remove_playlist(id=prov_item_id)
+            else:
+                raise ValueError("Unknown media type: %s", str(media_type))
+        except Exception as e:
+            self.logger.error(e)
+            return False
+        return True
+
+    async def add_playlist_tracks(self, prov_playlist_id: str, prov_track_ids: list[str]) -> None:
+        """Add track(s) to playlist."""
+        for id in prov_track_ids:
+            self.api.add_to_playlist(track_id=id, playlist_id=prov_playlist_id)
+
+    async def remove_playlist_tracks( self, prov_playlist_id: str, positions_to_remove: tuple[int, ...] ) -> None:
+        """Remove track(s) from playlist."""
+        first = min(positions_to_remove)
+        trackdata = self.api.playlist_track_page(id=prov_playlist_id, start=first, nb=max(positions_to_remove))
+        tracklist = self._parse_tracks(trackdata)
+        for track in [tracklist[x-first] for x in positions_to_remove]:
+            self.api.remove_from_playlist(track_id=track.item_id, playlist_id=prov_playlist_id)
+
+    async def create_playlist(self, name: str) -> Playlist:
+        """Create a new playlist on provider with given name."""
+        await self.api.create_playlist(title=name)
+
+    async def get_stream_details(self, item_id: str, media_type: MediaType) -> StreamDetails:
+        """Get streamdetails for a track/radio."""
+        # Get stream details for a track or radio.
+        # Implementing this method is MANDATORY to allow playback.
+        # The StreamDetails contain info how Music Assistant can play the track.
+        # item_id will always be a track or radio id. Later, when/if MA supports
+        # podcasts or audiobooks, this may as well be an episode or chapter id.
+        # You should return a StreamDetails object here with the info as accurate as possible
+        # to allow Music Assistant to process the audio using ffmpeg.
+        data = await self.api.track(id=item_id)
+
+        urldata = data['FALLBACK'] if 'FALLBACK' in data else data # original refreezer uses fallback by default too (seems like newer versions are stored in 'fallback', and outdated versions stop to function after a while)
+        url, encrypted = await self.api.get_track_url(track_id=urldata['SNG_ID'], track_token=urldata['TRACK_TOKEN'], md5origin=urldata['MD5_ORIGIN'], media_version=urldata['MEDIA_VERSION'], quality=self.config.get_value(CONF_REFREEZER_QUALITY))
+        if url == None:
+            raise UnplayableMediaError(f'Could not fetch url for track')
+        self.logger.debug(f'Found this url (encrypted={encrypted}): {url}')
+
+        details = self._parse_track_details(data)
+        details.data['url'] = url
+        details.data['encrypted'] = encrypted
+        details.provider = self.lookup_key
+        details.stream_type = StreamType.CUSTOM
+        details.can_seek=True
+        details.allow_seek=True
+        return details
+
+
+    def _md5(self, data, data_type="ascii"):
+        md5sum = md5()
+        md5sum.update(data.encode(data_type))
+        return md5sum.hexdigest()
+    def get_blowfish_key(self, track_id):
+        """Get blowfish key to decrypt a chunk of a track."""
+        # secret = 'jo6aey6haid2Teih' # refreezer official key
+        secret = 'g4el58wc0zvf9na1' # deezer mass-implementation key (found in app_var(5))
+        self.logger.info(f'got secret: {secret}')
+        id_md5 = self._md5(track_id)
+        return "".join(chr(ord(id_md5[i]) ^ ord(id_md5[i + 16]) ^ ord(secret[i])) for i in range(16))
+    def decrypt_chunk(self, chunk, blowfish_key):
+        """Decrypt a given chunk using the blow fish key."""
+        cipher = Blowfish.new(
+            blowfish_key.encode("ascii"),
+            Blowfish.MODE_CBC,
+            b"\x00\x01\x02\x03\x04\x05\x06\x07",
+        )
+        return cipher.decrypt(chunk)
+
+    # seb NOTE: a very basic encryption is used for the stream, as the music of a sample url is well-audible despite noise.
+    # Found these options to decrypt:
+    # 1. deezer mass provider
+    #   + we use the same url to call, so should process encryption data similarly
+    #   - uses stupid app_vars, so have to take it out again
+    # 2. refreezer DeezerDecryptor.
+    #   + seems simple enough
+    #   + might match encryption pattern (weak)
+    #   - not used by refreezer except when using the 'streaming server' and other offline tracks functionalities
+    # 3. refreezer 'generateTrackUrl'
+    #   + related to 2
+    #   - only used for offline track functionality?
+    #   - only call in 'getTrackUrl', with comment: "Deprecated legacy url generation. Seems unavailable now."
+    async def get_audio_stream(self, streamdetails: StreamDetails, seek_position: int=0) -> AsyncGenerator[bytes, None]:
+        """ Return the (custom) audio stream for the provider item. """
+        blowfish_key = self.get_blowfish_key(streamdetails.data['track_id'])
+        chunk_index = 0
+        timeout = ClientTimeout(total=0, connect=30, sock_read=600)
+        # headers = self.api.get_headers()
+        headers = {}
+        # if seek_position and streamdetails.size:
+        #     chunk_count = ceil(streamdetails.size / 2048)
+        #     chunk_index = int(chunk_count / streamdetails.duration) * seek_position
+        #     skip_bytes = chunk_index * 2048
+        #     headers["Range"] = f"bytes={skip_bytes}-"
+
+        # NOTE: Seek with using the Range header is not working properly
+        # causing malformed audio so this is a temporary patch
+        # by just skipping chunks
+        if seek_position and streamdetails.size:
+            chunk_count = ceil(streamdetails.size / 2048)
+            skip_chunks = int(chunk_count / streamdetails.duration) * seek_position
+        else:
+            skip_chunks = 0
+
+        buffer = bytearray()
+        streamdetails.data['start_ts'] = utc_timestamp()
+        streamdetails.data['stream_id'] = uuid.uuid1()
+        # self.mass.create_task(self.gw_client.log_listen(next_track=streamdetails.item_id))
+        async with self.api.session.get(streamdetails.data['url'], headers=headers, timeout=timeout) as resp:
+            self.logger.info(f'Got stream status {resp.status}, reason="{resp.reason}"')
+            self.logger.info(f'Resulting stream is of type "{resp.content_type}". Found headers: "{resp.headers}"')
+            async for chunk in resp.content.iter_chunked(2048):
+                buffer += chunk
+                if len(buffer) >= 2048:
+                    if chunk_index >= skip_chunks or chunk_index == 0:
+                        if chunk_index % 3 > 0:
+                            yield bytes(buffer[:2048])
+                        else:
+                            yield self.decrypt_chunk(bytes(buffer[:2048]), blowfish_key)
+
+                    chunk_index += 1
+                    del buffer[:2048]
+        yield bytes(buffer)
+
+
+    async def on_played( self, media_type: MediaType, prov_item_id: str, fully_played: bool, position: int, media_item: MediaItemType, is_playing: bool=False) -> None:
+        """Handle callback when a track is playing (called every 30s), has been fully played, or has been stopped/skipped."""
+        if fully_played: # seb NOTE: refreezer officially logs when >75% of a track has been played.
+            self.api.log_listen(id=prov_item_id)
+
+
+    def _parse_albums(self, data: Dict) -> list[Album]:
+        return [self._parse_album(elem) for elem in data]
+
+    def _parse_album(self, data: Dict) -> list[Album]:
+        artists = [self._parse_artist(elem) for elem in data['ARTISTS']] if 'ARTISTS' in data else []
+        if 'ARTISTS' not in data and 'ART_NAME' in data:
+            artists.append(self._parse_artist(data))
+
+        album = Album(
+            item_id=data['ALB_ID'], # seb TODO: tostring? Is now int
+            provider=self.lookup_key,
+            name=data['ALB_TITLE'],
+            year=int(data['DIGITAL_RELEASE_DATE'].split('-',1)[0]) if 'DIGITAL_RELEASE_DATE' in data else (int(data['PHYSICAL_RELEASE_DATE'].split('-',1)[0]) if 'PHYSICAL_RELEASE_DATE' in data else None),
+            # version= # seb TODO: spotify plugin provides this
+            provider_mappings={
+                ProviderMapping(
+                    item_id=data['ALB_ID'],
+                    provider_domain=self.domain,
+                    provider_instance=self.instance_id,
+                    available=True,
+                    # audio_format=AudioFormat(content_type=ContentType.OGG, bit_rate=320),
+                )
+            },
+            artists=artists,
+            album_type = AlbumType.SINGLE if 'TYPE' in data and str(data['TYPE']) == '0' else AlbumType.ALBUM
+        )
+        if 'ALB_PICTURE' in data and data['ALB_PICTURE'] != None:
+            details = DeezerAPI.ImageDetails.from_private_string(data['ALB_PICTURE']) # seb TODO: used to be: 'from_private_json' (but that does not work from search-origin)
+            album.metadata.images = [
+                MediaItemImage(
+                    type=ImageType.THUMB,
+                    path=details.thumb_url, # seb TODO: can also provide full url? Maybe both?
+                    provider=self.lookup_key,
+                    remotely_accessible=True
+                )
+            ]
+        return album
+
+    def _parse_artists(self, data: Dict) -> list[Artist]:
+        return [self._parse_artist(elem) for elem in data]
+
+    def _parse_artist(self, data: Dict) -> Artist:
+        artist = Artist(
+            item_id=data['ART_ID'],
+            provider=self.lookup_key,
+            name=data['ART_NAME'],
+            provider_mappings={
+                ProviderMapping( # seb TODO: what is this stuff?
+                    item_id=data['ART_ID'], # seb TODO: tostring? is now int
+                    provider_domain=self.domain,
+                    provider_instance=self.instance_id,
+                    available=True,
+                )
+            },
+        )
+        if 'ART_PICTURE' in data and data['ART_PICTURE'] != None:
+            details = DeezerAPI.ImageDetails.from_private_string(data['ART_PICTURE']) # seb TODO: used to be: 'from_private_json' (but that does not work from search-origin)
+            artist.metadata.images = [
+                MediaItemImage(
+                    type=ImageType.THUMB,
+                    path=details.thumb_url, # seb TODO: can also provide full url? Maybe both?
+                    provider=self.lookup_key,
+                    remotely_accessible=True
+                )
+            ]
+        # seb TODO: no use for 'highlight' data provided by deezer API?
+        return artist
+
+    def _parse_tracks(self, data: Dict) -> list[Track]:
+        return [self._parse_track(elem) for elem in data]
+
+    def _parse_track(self, data: Dict) -> Track:
+        version = data.get('VERSION')
+        title = data['SNG_TITLE'] if version is None else f"{data['SNG_TITLE']}"
+        track = Track(
+            item_id=data['SNG_ID'], # seb TODO: tostring? is now int
+            provider=self.lookup_key,
+            name=title,
+            version=version,
+            duration=int(data['DURATION']),
+            provider_mappings={
+                ProviderMapping(
+                    item_id=data['SNG_ID'],
+                    provider_domain=self.domain,
+                    provider_instance=self.instance_id,
+                    available=True,
+                    # audio_format=AudioFormat(content_type=ContentType.OGG, bit_rate=320),
+                ),
+            },
+            disc_number=int(data['DISK_NUMBER'] if 'DISK_NUMBER' in data and data['DISK_NUMBER'] != None else 0),
+            track_number=int(data['TRACK_NUMBER'] if 'TRACK_NUMBER' in data and data['TRACK_NUMBER'] != None else 0),
+            artists=self._parse_artists(data['ARTISTS'] if 'ARTISTS' in data else [data]),
+            album=self._parse_album(data)
+        )
+        track.metadata.explicit = 'EXPLICIT_LYRICS' in data and data['EXPLICIT_LYRICS'] == 1
+        if 'ALB_PICTURE' in data:
+            details = DeezerAPI.ImageDetails.from_private_string(data['ALB_PICTURE']) # seb TODO: used to be: 'from_private_json' (but that does not work from search-origin)
+            track.metadata.images = [
+                MediaItemImage(
+                    type=ImageType.THUMB,
+                    path=details.thumb_url, # seb TODO: can also provide full url? Maybe both?
+                    provider=self.lookup_key,
+                    remotely_accessible=True
+                )
+            ]
+        return track
+
+    def _parse_playlists(self, data: Dict) -> list[Playlist]:
+        return [self._parse_playlist(elem) for elem in data]
+
+    def _parse_playlist(self, data: Dict[str, Any]) -> Playlist:
+        is_editable = data['PARENT_USER_ID'] == self.api.user_id if data['PARENT_USER_ID'] != None else False # if id == None --> not made by us
+        playlist = Playlist(
+            item_id = data['PLAYLIST_ID'],
+            provider=self.instance_id if is_editable else self.lookup_key,
+            name=data['TITLE'],
+            owner=data.get('PARENT_USERNAME', ''),
+            provider_mappings={
+                ProviderMapping(
+                    item_id=data['PLAYLIST_ID'],
+                    provider_domain=self.domain,
+                    provider_instance=self.instance_id,
+                    available=True,
+                )
+            },
+            is_editable=is_editable,
+        )
+        details = DeezerAPI.ImageDetails.from_private_string(data['PLAYLIST_PICTURE']) # seb TODO: used to be: 'from_private_json' (but that does not work from search-origin)
+        playlist.metadata.images = [
+            MediaItemImage(
+                type=ImageType.THUMB,
+                path=details.thumb_url, # seb TODO: can also provide full url? Maybe both?
+                provider=self.lookup_key,
+                remotely_accessible=True
+            )
+        ]
+        # playlist.cache_checksum = str(playlist_obj["snapshot_id"])
+        # seb NOTE: no checksums provided by Deezer
+        return playlist
+
+    def _parse_track_details(self, data: Dict) -> StreamDetails:
+        content_type = ContentType.UNKNOWN # let FFMPEG guess the contenttype
+        bitrate = None
+        quality = self.config.get_value(CONF_REFREEZER_QUALITY)
+        size = None
+        if f'FILESIZE_{quality}' in data and data[f'FILESIZE_{quality}'] != '0':
+            size = data[f'FILESIZE_{quality}']
+            c = quality.split('_')
+            content_type = ContentType.try_parse(c[0])
+            bitrate = int(c[1] if len(c) >= 2 else None)
+        else:
+            for cat in ['AAC_64', 'MP3_64', 'MP3_128', 'MP3_256', 'MP3_320', 'FLAC']:
+                if data[f'FILESIZE_{cat}'] != "0":
+                    size = data[f'FILESIZE_{cat}']
+                    c = cat.split('_')
+                    content_type = ContentType.try_parse(c[0])
+                    bitrate = int(c[1] if len(c) >= 2 else None)
+
+        self.logger.debug(f'Using format details for item id {data["SNG_ID"]}: {content_type} @ {bitrate} kbits')
+        return StreamDetails(
+            item_id=data['SNG_ID'],
+            provider=self.lookup_key,
+            audio_format=AudioFormat(content_type=content_type, bit_rate=bitrate), # seb TODO: have to set bit_depth (default 16) and sample_rate (default 44100) too?
+            stream_type=StreamType.CUSTOM, # seb NOTE: used to be HTTP
+            duration=int(data['DURATION']),
+            data={'track_id': data['SNG_ID']}, # seb NOTE: cann pass in data here, passed on to the get_audio_stream funciton.
+            size=int(size) if size else None,
+            can_seek=True,
+            allow_seek=True,
+        )
diff --git a/music_assistant/providers/refreezer/manifest.json b/music_assistant/providers/refreezer/manifest.json
new file mode 100644
index 00000000..6d5ace3f
--- /dev/null
+++ b/music_assistant/providers/refreezer/manifest.json
@@ -0,0 +1,10 @@
+{
+  "type": "music",
+  "domain": "refreezer",
+  "name": "ReFreezer",
+  "description": "An alternative Deezer music streaming & downloading client, based on Freezer",
+  "documentation": "https://github.com/DJDoubleD/refreezer",
+  "codeowners": ["custom"],
+  "requirements": [],
+  "multi_instance": true
+}
