diff --git a/music_assistant/providers/refreezer/__init__.py b/music_assistant/providers/refreezer/__init__.py
new file mode 100644
index 00000000..abe414a4
--- /dev/null
+++ b/music_assistant/providers/refreezer/__init__.py
@@ -0,0 +1,924 @@
+from __future__ import annotations
+
+from collections.abc import AsyncGenerator, Sequence
+from typing import TYPE_CHECKING
+
+from hashlib import md5
+import urllib.parse
+
+from aiohttp import ClientSession, ClientTimeout, ClientResponse
+
+from music_assistant_models.config_entries import ConfigEntry, ConfigValueType, ProviderConfig
+from music_assistant_models.enums import (
+    AlbumType,
+    ConfigEntryType,
+    ContentType,
+    ExternalID,
+    ImageType,
+    MediaType,
+    ProviderFeature,
+    StreamType,
+)
+from music_assistant_models.media_items import (
+    Album,
+    Artist,
+    AudioFormat,
+    ItemMapping,
+    MediaItemImage,
+    MediaItemMetadata,
+    MediaItemType,
+    Playlist,
+    ProviderMapping,
+    RecommendationFolder,
+    SearchResults,
+    Track,
+)
+from music_assistant_models.provider import ProviderManifest
+from music_assistant_models.streamdetails import StreamDetails
+
+from music_assistant import MusicAssistant
+from music_assistant.helpers.app_vars import app_var
+from music_assistant.helpers.auth import AuthenticationHelper
+from music_assistant.helpers.datetime import utc_timestamp
+from music_assistant.models import ProviderInstanceType
+from music_assistant.models.music_provider import MusicProvider
+
+import random
+import json
+import logging
+from typing import Optional, Dict, Any, List
+
+
+
+SUPPORTED_FEATURES = {
+    ProviderFeature.LIBRARY_ARTISTS,
+    ProviderFeature.LIBRARY_ALBUMS,
+    ProviderFeature.LIBRARY_TRACKS,
+    ProviderFeature.LIBRARY_PLAYLISTS,
+    ProviderFeature.LIBRARY_ALBUMS_EDIT,
+    ProviderFeature.LIBRARY_TRACKS_EDIT,
+    ProviderFeature.LIBRARY_ARTISTS_EDIT,
+    ProviderFeature.LIBRARY_PLAYLISTS_EDIT,
+    ProviderFeature.ALBUM_METADATA,
+    ProviderFeature.TRACK_METADATA,
+    ProviderFeature.ARTIST_METADATA,
+    ProviderFeature.ARTIST_ALBUMS,
+    ProviderFeature.ARTIST_TOPTRACKS,
+    ProviderFeature.BROWSE,
+    ProviderFeature.SEARCH,
+    ProviderFeature.PLAYLIST_TRACKS_EDIT,
+    ProviderFeature.PLAYLIST_CREATE,
+    ProviderFeature.RECOMMENDATIONS,
+    ProviderFeature.SIMILAR_TRACKS,
+}
+
+CONF_REFREEZER_MAIL = "refreezer_mail"
+CONF_REFREEZER_PASS = "refreezer_pass"
+CONF_REFREEZER_ARL = "refreezer_arl"
+
+CONF_ACTION_AUTH = "arl_token"
+
+
+DEFAULT_HEADERS = {
+    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36',
+    'X-User-IP': '1.1.1.1',
+    'x-deezer-client-ip': '1.1.1.1',
+    'Accept': '*/*'
+}
+
+class CookieManager:
+    '''handles obtaining session cookies from deezer'''
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self.cookie_header = {};
+        self.cookies = {};
+
+    def update_cookie(self, r: ClientResponse):
+        '''updates cookie info'''
+        if not any(r.headers['set-cookie']):
+            return
+        setcookies = r.headers.getall('set-cookie')
+        for setc in setcookies:
+            cookies = setc.split(';')
+            for c in cookies:
+                self._set_cookie(c)
+        self.cookie_header['cookie'] = self._generate_cookie_header();
+
+    def _set_cookie(self, cookie: str):
+        if (not cookie) or (not '=' in cookie):
+            return
+        key, value = [s.strip() for s in cookie.split('=', 1)]
+        if key.lower() in ['path', 'expires', 'domain', 'samesite']:
+            return # not interesting
+        # if we get here, this is an interesting cookie - keep it
+        self.cookies[key] = value
+
+    def _generate_cookie_header(self):
+        return ';'.join((f'{k}={v}'for k, v in self.cookies.items()))
+
+
+async def get_arl_by_email_and_password(email: str, password: str, http_session: ClientSession) -> str:
+    '''Automatically returns ARL token by using the deezer API'''
+    cookie_manager = CookieManager()
+    # remove any existing cookies from http_session
+    http_session.cookie_jar.clear()
+
+    # get initial cookies (sid) from empty getUser call
+    response = await http_session.get('https://www.deezer.com/ajax/gw-light.php?method=deezer.getUserData&input=3&api_version=1.0&api_token=null')
+    cookie_manager.update_cookie(response)
+    print(f'initial cookies success: {cookie_manager.cookie_header}')
+    http_session.cookie_jar.clear()
+
+    # Try to get AccessToken by login with email & password, which sets authentication cookies
+    token = await _get_accesstoken(email, password, cookie_manager, http_session)
+    if (token == None):
+        return ''
+
+    # get arl
+    request_headers = DEFAULT_HEADERS | cookie_manager.cookie_header
+    response = await http_session.get('https://www.deezer.com/ajax/gw-light.php?method=user.getArl&input=3&api_version=1.0&api_token=null', headers=request_headers)
+
+    data = await response.json()
+    return data['results']
+
+async def _get_accesstoken(email: str, password: str, cookie_manager: CookieManager, http_session: ClientSession):
+    # music-assistant
+    # DEEZER_APP_ID = '600344'
+    # DEEZER_APP_S = '1e79e9ecb000010147b7e31c21794b74'
+    # Spleeter, Deezloader creds (maybe)
+    # DEEZER_APP_ID = '243762'
+    # DEEZER_APP_S = 'a5cf06353c1d7083e2fca5a6885a1a74'
+    # Deezer API (same as streamrip & Orpheus)
+    DEEZER_APP_ID = '447462'
+    DEEZER_APP_S = 'a83bf7f38ad2f137e444727cfc3775cf'
+    # LastFM API original freezer
+    # LASTFM_API_ID = 'b6ab5ae967bcd8b10b23f68f42493829'
+    # LASTFM_API_S = '861b0dff9a8a574bec747f9dab8b82bf'
+
+    request_headers = DEFAULT_HEADERS | cookie_manager.cookie_header
+    
+    hashed_pass = md5(password.encode('utf-8')).hexdigest()
+    hashed_params = md5(f'{DEEZER_APP_ID}{email}{hashed_pass}{DEEZER_APP_S}'.encode('utf-8')).hexdigest()
+
+    params = {'app_id': DEEZER_APP_ID, 'login': email, 'password': hashed_pass, 'hash': hashed_params}
+    url = f'https://connect.deezer.com/oauth/user_auth.php?{urllib.parse.urlencode(params)}'
+
+    response = await http_session.get(url, headers=request_headers)
+    cookie_manager.update_cookie(response)
+    json = await response.json()
+    if 'access_token' in json:
+        return json['access_token']
+    elif 'error' in json and any(json['error']):
+        raise ValueError(f'got an error of type \'{json["error"]["type"]}\', message: \'{json["error"]["message"]}\'')
+    return None
+
+
+async def setup(
+    mass: MusicAssistant, manifest: ProviderManifest, config: ProviderConfig
+) -> ProviderInstanceType:
+    """Initialize provider(instance) with given configuration."""
+    return ReFreezerProvider(mass, manifest, config)
+
+
+async def get_config_entries(
+    mass: MusicAssistant,
+    instance_id: str | None = None,
+    action: str | None = None,
+    values: dict[str, ConfigValueType] | None = None,
+) -> tuple[ConfigEntry, ...]:
+    """
+    Return Config entries to setup this provider.
+
+    instance_id: id of an existing provider instance (None if new instance setup).
+    action: [optional] action key called from config entries UI.
+    values: the (intermediate) raw values for config entries sent with the action.
+    """
+    if action == CONF_ACTION_AUTH:
+        values[CONF_REFREEZER_ARL] = await get_arl_by_email_and_password(values[CONF_REFREEZER_MAIL], values[CONF_REFREEZER_PASS], mass.http_session)
+
+    return (
+        ConfigEntry(
+            key=CONF_REFREEZER_MAIL,
+            type=ConfigEntryType.STRING,
+            label="email",
+            required=True,
+            description="Deezer email",
+            value=values.get(CONF_REFREEZER_MAIL) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_REFREEZER_PASS,
+            type=ConfigEntryType.SECURE_STRING,
+            label="password",
+            required=True,
+            description="Deezer password",
+            value=values.get(CONF_REFREEZER_PASS) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_REFREEZER_ARL,
+            type=ConfigEntryType.SECURE_STRING,
+            label="Get ARL token",
+            required=True,
+            action=CONF_ACTION_AUTH,
+            description="Deezer ARL token",
+            value=values.get(CONF_REFREEZER_ARL) if values else None,
+        ),
+    )
+
+class DeezerAPI:
+    class Settings:
+        defaults = {
+            'deezer_language': 'en',
+            'deezer_country': 'US'
+        }
+        def __init__(**kwargs):
+            '''populate settings using passed kwargs, default value otherwise'''
+            pick = lambda name: kwargs[name] if name in kwargs else defaults[name]
+            self.deezer_language: str = pick('deezer_language')
+            self.deezer_country: str = pick('deezer_country')
+
+    class ImageDetails:
+        def __init__(self, full_url: str, thumb_url: str, type: str, image_hash: str):
+            self.full_url = full_url
+            self.thumb_url = thumb_url
+            self.type = type
+            self.image_hash = image_hash
+
+        def from_private_string(image_hash: str, type: str = 'cover') -> ImageDetails:
+            return ImageDetails(full_url=custom_url(image_hash, type, 1000, 1000, 80), thumb_url=custom_url(image_hash, type, 140, 140, 80), type=type, image_hash=image_hash)
+
+        def from_private_json(data: Dict) -> ImageDetails:
+            return ImageDetails.from_private_string(data['MD5'] if 'MD5' in data and data['MD5'] != None else data['md5'], data['TYPE'] if 'TYPE' in data and data['TYPE'] != None else data['type'])
+
+        def custom_url(image_hash: str, type: str, height: int, width: int, quality: int) -> str:
+            return f'https://e-cdns-images.dzcdn.net/images/{type}/{image_hash}/{height}x{width}-000000-{quality}-0-0.jpg'
+            
+        def custom_url(self, height: int, width: int, quality: int) -> str:
+            return ImageDetail.custom_url(self.image_hash, self.type, height, width, quality)
+
+
+    user_agent = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36'
+
+    def __init__(self, arl: Optional[str] = None, settings: Optional[Settings] = None):
+        self.arl = arl
+        self.token: Optional[str] = None
+        self.license_token: Optional[str] = None
+        self.user_id: Optional[str] = None
+        self.user_name: Optional[str] = None
+        self.favorites_playlist_id: Optional[str] = None
+        self.sid: Optional[str] = None
+        self._authorizing: Optional[asyncio.Future] = None
+        self.settings: Settings = settings or Settings()
+
+    def get_headers(self) -> Dict[str, str]:
+        lang = f'{self.settings.deezer_language}-{self.settings.deezer_country}'
+        cookie = f'arl={self.arl}' if self.sid == None else f'arl={self.arl}; sid={self.sid}'
+        return {
+            'User-Agent': self.user_agent,
+            'Content-Language': lang,
+            'Content-Type': 'text/plain;charset=UTF-8',
+            'Accept': ''*/*'',
+            'Accept-Charset': 'utf-8,ISO-8859-1;q=0.7,*;q=0.3',
+            'Accept-Language': f'{lang},{self.settings.deezer_language};q=0.9,en-US;q=0.8,en;q=0.7',
+            'Connection': 'keep-alive',
+            'Cookie': cookie,
+        }
+
+    async def call_gw_api(self, method: str, params: Optional[Dict] = None, gateway_input: Optional[str] = None) -> Dict[str, Any]:
+        url = 'https://www.deezer.com/ajax/gw-light.php'
+        query = {
+            'api_version': '1.0',
+            'api_token': self.token,
+            'input': '3',
+            'method': method,
+            'cid': str(random.randint(0, 1_000_000_000)),
+        }
+        if gateway_input:
+            query['gateway_input'] = gateway_input
+
+        headers = self.get_headers()
+        async with aiohttp.ClientSession() as session: # seb TODO: Could reuse hass-wide session?
+            async with session.post(url, headers=headers, params=query, data=json.dumps(params or {})) as resp:
+                body = await resp.json()
+
+                # Grab SID cookie
+                if method == 'deezer.getUserData':
+                    set_cookie = resp.headers.get('Set-Cookie', '')
+                    for cookie in set_cookie.split('';''):
+                        if cookie.strip().startswith('sid=''):
+                            self.sid = cookie.strip().split(''='', 1)[1]
+
+                # In case of error 'Invalid CSRF token' retrieve new one and retry the same call
+                # Except for 'deezer.getUserData' method, which would cause infinite loop
+                if ( 'error' in body and 'VALID_TOKEN_REQUIRED' in body['error'] and method != 'deezer.getUserData' and await self.raw_authorize() ):
+                    return await self.call_gw_api(method, params=params, gateway_input=gateway_input)
+                return body
+
+      async def call_public_api(self, path: str) -> Dict[str, Any]:
+            async with aiohttp.ClientSession() as session:
+                async with session.get(f'https://api.deezer.com/{path}'') as resp:
+                    return await resp.json()
+
+        async def get_json_web_token(self) -> str:
+            url = 'https://auth.deezer.com/login/arl'
+            params = {'jo': 'p', 'rto': 'c', 'i': 'c'}
+            headers = self.get_headers()
+            async with aiohttp.ClientSession() as session:
+                async with session.post(url, headers=headers, params=params) as resp:
+                    data = await resp.json()
+                    return data.get('jwt', '')
+
+        async def call_pipe_api(self, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
+            jwt_token = await self.get_json_web_token()
+            headers = self.get_headers()
+            headers['Authorization'] = f'Bearer {jwt_token}'
+            headers['Content-Type'] = 'application/json'
+            url = 'https://pipe.deezer.com/api/'
+            async with aiohttp.ClientSession() as session:
+                async with session.post(url, headers=headers, data=json.dumps(params or {})) as resp:
+                    return await resp.json()
+
+    async def authorize(self) -> bool:
+        if self._authorizing is None:
+            self._authorizing = asyncio.ensure_future(self.raw_authorize())
+            result = await self._authorizing
+            self._authorizing = None
+            return result
+        return await self._authorizing
+
+    async def raw_authorize(self) -> bool:
+        try:
+            data = await self.call_gw_api("deezer.getUserData")
+            user = data.get("results", {}).get("USER", {})
+            if user.get("USER_ID") == 0:
+                return False
+
+            self.token = data["results"].get("checkForm")
+            self.user_id = str(user.get("USER_ID", ""))
+            self.user_name = user.get("BLOG_NAME")
+            self.favorites_playlist_id = user.get("LOVEDTRACKS_ID")
+            self.license_token = user.get("OPTIONS", {}).get("license_token")
+            return True
+        except Exception as e:
+            log.exception("Login Error (D): %s", e) # seb TODO: music assistant logging
+            return False
+
+# seb TODO:
+# parseLink
+# checkAvailability
+
+
+    async def search(self, query: str) -> Dict[str, Any]:
+        data = await self.call_gw_api('deezer.pageSearch', params={'nb': 128, 'query': query, 'start': 0}) # seb TODO: respect music assistant limit?
+        return data.get('results', {})
+
+    async def track(self, id: str) -> Dict[str, Any]:
+        data = await self.call_gw_api('song.getListData', params={'sng_ids': [id]})
+        return data['results']['data'][0]
+
+    async def album(self, id: str) -> Dict[str, Any]:
+        data = await self.call_gw_api('deezer.pageAlbum', params={'alb_id': id, 'header': True, 'lang': self.settings.deezer_language})
+        return data['results']
+
+    async def artist(self, id: str) -> Dict[str, Any]:
+        data = await self.call_gw_api('deezer.pageArtist', params={'art_id': id, 'lang': self.settings.deezer_language})
+        return data['results']
+
+# seb TODO:
+# playlistTracksPage
+
+    async def playlist(self, id: str, nb: int = 100) -> Dict[str, Any]:
+        data = await self.call_gw_api('deezer.pagePlaylist', params={'playlist_id': id, 'lang': self.settings.deezer_language, 'nb': nb, 'tags': True, 'start': 0})
+        return data['results']
+
+    async def full_playlist(self, id: str) -> Dict[str, Any]:
+        return await self.playlist(id, nb=100000)
+
+    async def add_favorite_track(self, id: str):
+        await self.call_gw_api('favorite_song.add', params={'SNG_ID': id})
+
+# seb TODO:
+# addFavoriteAlbum
+# addFavoriteArtist
+# removeArtist
+
+    async def dislike_track(self, id: str):
+        await self.call_gw_api('favorite_dislike.add', params={'ID': id, 'TYPE': 'song'})
+
+    async def add_to_playlist(self, track_id: str, playlist_id: str, offset: int = -1):
+        await self.call_gw_api('playlist.addSongs', params={'offset': offset, 'playlist_id': playlist_id, 'songs': [[track_id, 0]]})
+
+    async def remove_from_playlist(self, track_id: str, playlist_id: str):
+        await self.call_gw_api('playlist.deleteSongs', params={'playlist_id': playlist_id, 'songs': [[track_id, 0]]})
+
+    async def create_playlist(self, title: str, description: str = '', status: int = 1, track_ids: List[str] = []) -> str:
+        songs = [[int(tid), idx] for idx, tid in enumerate(track_ids)]
+        data = await self.call_gw_api('playlist.create', params={'title': title, 'description': description, 'songs': songs, 'status': status})
+        return str(data['results'])
+
+    async def delete_playlist(self, id: str):
+        await self.call_gw_api('playlist.delete', params={'playlist_id': id})
+
+    async def update_playlist(self, id: str, title: str, description: str, status: int = 1):
+        await self.call_gw_api('playlist.update', params={'description': description, 'title': title, 'playlist_id': int(id), 'status': status, 'songs': []})
+
+
+# seb TODO:
+# getPlaylists
+# getFavoriteTrackIDs
+# getAlbums
+# removeAlbum
+# removeFavorite
+# getArtists
+# lyrics
+# lyricsLegacy
+# lyricsFull
+# smartTrackList
+# flow
+# homePage
+# logListen
+# getChannel
+# addPlaylist
+# removePlaylist
+# discographyPage
+# searchSuggestions
+# smartRadio
+# libraryShuffle
+# playMix
+# allShowEpisodes
+
+
+
+class ReFreezerProvider(MusicProvider):
+    """ReFreezer provider support."""
+
+    api: DeezerAPI
+
+    async def handle_async_init(self, arl: Optional[str] = None, settings: Optional[DeezerAPI.Settings] = None) -> None:
+        self.api = DeezerAPI(arl=arl, settings=settings)
+
+    @property
+    def supported_features(self) -> set[ProviderFeature]:
+        """Return the features supported by this Provider."""
+        return SUPPORTED_FEATURES
+
+    # @property
+    # def is_streaming_provider(self) -> bool:
+    #     """
+    #     Return True if the provider is a streaming provider.
+
+    #     This literally means that the catalog is not the same as the library contents.
+    #     For local based providers (files, plex), the catalog is the same as the library content.
+    #     It also means that data is if this provider is NOT a streaming provider,
+    #     data cross instances is unique, the catalog and library differs per instance.
+
+    #     Setting this to True will only query one instance of the provider for search and lookups.
+    #     Setting this to False will query all instances of this provider for search and lookups.
+    #     """
+    #     # For streaming providers return True here but for local file based providers return False.
+    #     return True
+
+    async def search( self, search_query: str, media_types: list[MediaType], limit: int = 5, ) -> SearchResults:
+        """Perform search on musicprovider.
+
+        :param search_query: Search query.
+        :param media_types: A list of media_types to include.
+        :param limit: Number of items to return in the search (per type).
+        """
+        data = await callGwApi('deezer.pageSearch', params: {'nb': 128, 'query': query, 'start': 0})
+        # seb TODO: parse data into categories... Other side uses SearchResults
+
+        # OPTIONAL
+        # Will only be called if you reported the SEARCH feature in the supported_features.
+        # It allows searching your provider for media items.
+        # See the model for SearchResults for more information on what to return, but
+        # in general you should return a list of MediaItems for each media type.
+
+    async def get_library_artists(self) -> AsyncGenerator[Artist, None]:
+        """Retrieve library artists from the provider."""
+        # OPTIONAL
+        # Will only be called if you reported the LIBRARY_ARTISTS feature
+        # in the supported_features and you did not override the default sync method.
+        # It allows retrieving the library/favorite artists from your provider.
+        # Warning: Async generator:
+        # You should yield Artist objects for each artist in the library.
+        yield Artist(
+            # A simple example of an artist object,
+            # you should replace this with actual data from your provider.
+            # Explore the Artist model for all options and descriptions.
+            item_id="123",
+            provider=self.instance_id,
+            name="Artist Name",
+            provider_mappings={
+                ProviderMapping(
+                    # A provider mapping is used to provide details about this item on this provider
+                    # Music Assistant differentiates between domain and instance id to account for
+                    # multiple instances of the same provider.
+                    # The instance_id is auto generated by MA.
+                    item_id="123",
+                    provider_domain=self.domain,
+                    provider_instance=self.instance_id,
+                    # set 'available' to false if the item is (temporary) unavailable
+                    available=True,
+                    audio_format=AudioFormat(
+                        # provide details here about sample rate etc. if known
+                        content_type=ContentType.FLAC,
+                    ),
+                )
+            },
+        )
+
+    async def get_library_albums(self) -> AsyncGenerator[Album, None]:
+        """Retrieve library albums from the provider."""
+        # OPTIONAL
+        # Will only be called if you reported the LIBRARY_ALBUMS feature
+        # in the supported_features and you did not override the default sync method.
+        # It allows retrieving the library/favorite albums from your provider.
+        # Warning: Async generator:
+        # You should yield Album objects for each album in the library.
+
+        yield  # type: ignore[misc]
+
+    async def get_library_tracks(self) -> AsyncGenerator[Track, None]:
+        """Retrieve library tracks from the provider."""
+        # OPTIONAL
+        # Will only be called if you reported the LIBRARY_TRACKS feature
+        # in the supported_features and you did not override the default sync method.
+        # It allows retrieving the library/favorite tracks from your provider.
+        # Warning: Async generator:
+        # You should yield Track objects for each track in the library.
+        yield  # type: ignore[misc]
+
+    async def get_library_playlists(self) -> AsyncGenerator[Playlist, None]:
+        """Retrieve library/subscribed playlists from the provider."""
+        # OPTIONAL
+        # Will only be called if you reported the LIBRARY_PLAYLISTS feature
+        # in the supported_features and you did not override the default sync method.
+        # It allows retrieving the library/favorite playlists from your provider.
+        # Warning: Async generator:
+        # You should yield Playlist objects for each playlist in the library.
+        yield  # type: ignore[misc]
+
+    async def get_library_radios(self) -> AsyncGenerator[Radio, None]:
+        """Retrieve library/subscribed radio stations from the provider."""
+        # OPTIONAL
+        # Will only be called if you reported the LIBRARY_RADIOS feature
+        # in the supported_features and you did not override the default sync method.
+        # It allows retrieving the library/favorite radio stations from your provider.
+        # Warning: Async generator:
+        # You should yield Radio objects for each radio station in the library.
+        yield  # type: ignore[misc]
+
+    async def get_artist(self, prov_artist_id: str) -> Artist:  # type: ignore[empty-body]
+        """Get full artist details by id."""
+        data = await self.api.artist(id=prov_artist_id)
+        return self._parse_artist(data['results']['DATA'])
+
+    async def _parse_artist(data: Dict) -> Artist:
+         artist = Artist(
+            item_id=data['ART_ID'],
+            provider=self.lookup_key,
+            name=data['ART_NAME']
+            provider_mappings={
+                ProviderMapping( # seb TODO: what is this stuff?
+                    item_id=data['ART_ID'], # seb TODO: tostring? is now int
+                    provider_domain=self.domain,
+                    provider_instance=self.instance_id,
+                    url=None, # seb TODO: will this be ok? deezer does not provide a link
+                )
+            }
+        )
+        if 'ART_PICTURE' in data and data['ART_PICTURE'] != None:
+            details = DeezerAPI.ImageDetails.from_private_json(data['ART_PICTURE'])
+            artist.metadata.images = [
+                MediaItemImage(
+                    type=ImageType.THUMB,
+                    path=details.thumb_url, # seb TODO: can also provide full url? Maybe both?
+                    provider=self.lookup_key,
+                    remotely_accessible=True
+                )
+            ]
+        # seb TODO: no use for 'highlight' data provided by deezer API?
+        return artist
+
+
+    async def get_artist_albums(self, prov_artist_id: str) -> list[Album]:  # type: ignore[empty-body]
+        """Get a list of all albums for the given artist."""
+        data = await self.api.artist(id=prov_artist_id)
+        return self._parse_albums(data['results']['ALBUMS'])
+
+    async def _parse_albums(self, data: Dict) -> list[Album]:
+        return [self._parse_album(elem) for elem in data['data']] if any(data['data']) else []
+
+    async def _parse_album(self, data: Dict) -> list[Album]:
+        artists = [self._parse_artist(elem) for elem in data['ARTISTS']] if any(data['ARTISTS']) else []
+        if not any(data['ARTISTS']) and 'ART_NAME' in data:
+            artists.append(Artist(name=data['ART_NAME']))
+
+        type = 'single' if data['TYPE'] != None and str(data['TYPE']) == '0' else 'album'
+
+        album = Album(
+            item_id=data['ALB_ID'], # seb TODO: tostring? Is now int
+            provider=self.lookup_key,
+            name=data['ALB_TITLE'],
+            year=int((data['DIGITAL_RELEASE_DATE'] if data['DIGITAL_RELEASE_DATE'] else data['PHYSICAL_RELEASE_DATE']).split('-')[0]),
+            # version= # seb TODO: spotify plugin provides this
+            provider_mappings={
+                ProviderMapping(
+                    provider_domain=self.domain,
+                    provider_instance=self.instance_id,
+                    # audio_format=AudioFormat(content_type=ContentType.OGG, bit_rate=320),
+                    url=None, # seb TODO: will this be ok? deezer does not provide a link
+                )
+            }
+            artists=artists,
+            album_type=AlbumType(type)
+        )
+        if 'ALB_PICTURE' in data and data['ALB_PICTURE'] != None:
+            details = DeezerAPI.ImageDetails.from_private_json(data['ALB_PICTURE'])
+            album.metadata.images = [
+                MediaItemImage(
+                    type=ImageType.THUMB,
+                    path=details.thumb_url, # seb TODO: can also provide full url? Maybe both?
+                    provider=self.lookup_key,
+                    remotely_accessible=True
+                )
+            ]
+        
+
+    async def get_artist_toptracks(self, prov_artist_id: str) -> list[Track]:  # type: ignore[empty-body]
+        """Get a list of most popular tracks for the given artist."""
+        data = await self.api.artist(id=prov_artist_id)
+        return self._parse_tracks(data['results']['TOP'])
+
+    async def _parse_tracks(self, data: Dict) -> list[Track]:
+        return [self._parse_track(elem, album=album)] for elem in data['data'] if any(data['data']) else []
+
+    async def _parse_track(self, data: Dict) -> Track:
+        title = data['SNG_TITLE'] if not data['VERSION'] else f"{data['SNG_TITLE']} {data['VERSION']}"
+        track = Track(
+            item_id=data['SNG_ID'], # seb TODO: tostring? is now int
+            provider=self.lookup_key,
+            name=title,
+            version=data['VERSION']
+            duration=int(data['DURATION'])
+            provider_mappings={
+                ProviderMapping(
+                    item_id=data['SNG_ID'],
+                    provider_domain=self.domain,
+                    provider_instance=self.instance_id,
+                    # audio_format=AudioFormat(content_type=ContentType.OGG, bit_rate=320),
+                    url=None, # seb TODO: is this ok?
+                    available=True,
+                )
+            }
+            disc_number=int(data['DISK_NUMBER'] if 'DISK_NUMBER' in data and data['DISK_NUMBER'] != None else 0),
+            track_number=int(data['TRACK_NUMBER'] if 'TRACK_NUMBER' in data and data['TRACK_NUMBER'] != None else 0),
+            artists=self._parse_artists(data['ARTISTS'] if 'ARTISTS' in data else [data])
+            album=self._parse_album(data)
+        )
+        track.metadata.explicit = str(data['EXPLICIT_LYRICS']) == '1'
+        if 'ALB_PICTURE' in data:
+            details = DeezerAPI.ImageDetails.from_private_json(data['ALB_PICTURE'])
+            track.metadata.images = [
+                MediaItemImage(
+                    type=ImageType.THUMB,
+                    path=details.thumb_url, # seb TODO: can also provide full url? Maybe both?
+                    provider=self.lookup_key,
+                    remotely_accessible=True
+                )
+            ]
+        return track
+
+
+    async def get_album(self, prov_album_id: str) -> Album:  # type: ignore[empty-body]
+        """Get full album details by id."""
+        data = self.api.album(id=prov_album_id)
+        return self._parse_album(data['results']['DATA'])
+
+        # Get full details of a single Album.
+        # Mandatory only if you reported LIBRARY_ALBUMS in the supported_features.
+
+    async def get_track(self, prov_track_id: str) -> Track:  # type: ignore[empty-body]
+        """Get full track details by id."""
+        data = self.api.track(id=prov_track_id)
+        return self._parse_track(data['results']['data'][0])
+        # Get full details of a single Track.
+        # Mandatory only if you reported LIBRARY_TRACKS in the supported_features.
+
+    async def get_playlist(self, prov_playlist_id: str) -> Playlist:  # type: ignore[empty-body]
+        """Get full playlist details by id."""
+        # Get full details of a single Playlist.
+        # Mandatory only if you reported LIBRARY_PLAYLISTS in the supported
+
+    async def get_radio(self, prov_radio_id: str) -> Radio:  # type: ignore[empty-body]
+        """Get full radio details by id."""
+        # Get full details of a single Radio station.
+        # Mandatory only if you reported LIBRARY_RADIOS in the supported_features.
+
+    async def get_album_tracks( self, prov_album_id: str) -> list[Track]: # type: ignore[empty-body]
+        """Get album tracks for given album id."""
+        data = self.api.album(id=prov_album_id)
+        return self._parse_tracks(data['results']['SONGS'])
+
+
+    async def get_playlist_tracks(  # type: ignore[empty-body]
+        self,
+        prov_playlist_id: str,
+        page: int = 0,
+    ) -> list[Track]:
+        """Get all playlist tracks for given playlist id."""
+        # Get all tracks for a given playlist.
+        # Mandatory only if you reported LIBRARY_PLAYLISTS in the supported_features.
+
+    async def library_add(self, item: MediaItemType) -> bool:
+        """Add item to provider's library. Return true on success."""
+        # Add an item to your provider's library.
+        # This is only called if the provider supports the EDIT feature for the media type.
+        return True
+
+    async def library_remove(self, prov_item_id: str, media_type: MediaType) -> bool:
+        """Remove item from provider's library. Return true on success."""
+        # Remove an item from your provider's library.
+        # This is only called if the provider supports the EDIT feature for the media type.
+        return True
+
+    async def add_playlist_tracks(self, prov_playlist_id: str, prov_track_ids: list[str]) -> None:
+        """Add track(s) to playlist."""
+        # Add track(s) to a playlist.
+        # This is only called if the provider supports the PLAYLIST_TRACKS_EDIT feature.
+
+    async def remove_playlist_tracks(
+        self, prov_playlist_id: str, positions_to_remove: tuple[int, ...]
+    ) -> None:
+        """Remove track(s) from playlist."""
+        # Remove track(s) from a playlist.
+        # This is only called if the provider supports the PLAYLIST_TRACKS_EDIT feature.
+
+    async def create_playlist(self, name: str) -> Playlist:  # type: ignore[empty-body]
+        """Create a new playlist on provider with given name."""
+        # Create a new playlist on the provider.
+        # This is only called if the provider supports the PLAYLIST_CREATE feature.
+
+    async def get_similar_tracks(  # type: ignore[empty-body]
+        self, prov_track_id: str, limit: int = 25
+    ) -> list[Track]:
+        """Retrieve a dynamic list of similar tracks based on the provided track."""
+        # Get a list of similar tracks based on the provided track.
+        # This is only called if the provider supports the SIMILAR_TRACKS feature.
+
+    async def get_resume_position(self, item_id: str, media_type: MediaType) -> tuple[bool, int]:  # type: ignore[empty-body]
+        """
+        Get progress (resume point) details for the given Audiobook or Podcast episode.
+
+        This is a separate call from the regular get_item call to ensure the resume position
+        is always up-to-date and because a lot providers have this info present on a dedicated
+        endpoint.
+
+        Will be called right before playback starts to ensure the resume position is correct.
+
+        Returns a boolean with the fully_played status
+        and an integer with the resume position in ms.
+        """
+        # optional function to get the resume position of a audiobook or podcast episode
+        # only implement this if your provider supports providing this information
+
+    async def get_stream_details(self, item_id: str, media_type: MediaType) -> StreamDetails:
+        """Get streamdetails for a track/radio."""
+        # Get stream details for a track or radio.
+        # Implementing this method is MANDATORY to allow playback.
+        # The StreamDetails contain info how Music Assistant can play the track.
+        # item_id will always be a track or radio id. Later, when/if MA supports
+        # podcasts or audiobooks, this may as well be an episode or chapter id.
+        # You should return a StreamDetails object here with the info as accurate as possible
+        # to allow Music Assistant to process the audio using ffmpeg.
+        return StreamDetails(
+            provider=self.instance_id,
+            item_id=item_id,
+            audio_format=AudioFormat(
+                # provide details here about sample rate etc. if known
+                # set content type to unknown to let ffmpeg guess the codec/container
+                content_type=ContentType.UNKNOWN,
+            ),
+            media_type=MediaType.TRACK,
+            # streamtype defines how the stream is provided
+            # for most providers this will be HTTP but you can also use CUSTOM
+            # to provide a custom stream generator in get_audio_stream.
+            stream_type=StreamType.HTTP,
+            # explore the StreamDetails model and StreamType enum for more options
+            # but the above should be the mandatory fields to set.
+            allow_seek=True,
+            # set allow_seek to True if the stream may be seeked
+            can_seek=True,
+            # set can_seek to True if the stream supports seeking
+        )
+
+    async def get_audio_stream(
+        self, streamdetails: StreamDetails, seek_position: int = 0
+    ) -> AsyncGenerator[bytes, None]:
+        """
+        Return the (custom) audio stream for the provider item.
+
+        Will only be called when the stream_type is set to CUSTOM.
+        """
+        # this is an async generator that should yield raw audio bytes
+        # for the given streamdetails. You can use this to provide a custom
+        # stream generator for the audio stream. This is only called when the
+        # stream_type is set to CUSTOM in the get_stream_details method.
+        yield  # type: ignore[misc]
+
+    async def on_streamed(
+        self,
+        streamdetails: StreamDetails,
+    ) -> None:
+        """
+        Handle callback when given streamdetails completed streaming.
+
+        To get the number of seconds streamed, see streamdetails.seconds_streamed.
+        To get the number of seconds seeked/skipped, see streamdetails.seek_position.
+        Note that seconds_streamed is the total streamed seconds, so without seeked time.
+
+        NOTE: Due to internal and player buffering,
+        this may be called in advance of the actual completion.
+        """
+        # This is an OPTIONAL callback that is called when an item has been streamed.
+        # You can use this e.g. for playback reporting or statistics.
+
+    async def on_played(
+        self,
+        media_type: MediaType,
+        prov_item_id: str,
+        fully_played: bool,
+        position: int,
+        media_item: MediaItemType,
+        is_playing: bool = False,
+    ) -> None:
+        """
+        Handle callback when a (playable) media item has been played.
+
+        This is called by the Queue controller when;
+            - a track has been fully played
+            - a track has been stopped (or skipped) after being played
+            - every 30s when a track is playing
+
+        Fully played is True when the track has been played to the end.
+
+        Position is the last known position of the track in seconds, to sync resume state.
+        When fully_played is set to false and position is 0,
+        the user marked the item as unplayed in the UI.
+
+        is_playing is True when the track is currently playing.
+
+        media_item is the full media item details of the played/playing track.
+        """
+        # This is an OPTIONAL callback that is called when an item has been streamed.
+        # You can use this e.g. for playback reporting or statistics.
+
+    async def resolve_image(self, path: str) -> str | bytes:
+        """
+        Resolve an image from an image path.
+
+        This either returns (a generator to get) raw bytes of the image or
+        a string with an http(s) URL or local path that is accessible from the server.
+        """
+        # This is an OPTIONAL method that you can implement to resolve image paths.
+        # This is used to resolve image paths that are returned in the MediaItems.
+        # You can return a URL to an image or a generator that yields the raw bytes of the image.
+        # This will only be called when you set 'remotely_accessible'
+        # to false in a MediaItemImage object.
+        return path
+
+    async def browse(self, path: str) -> Sequence[MediaItemTypeOrItemMapping]:
+        """Browse this provider's items.
+
+        :param path: The path to browse, (e.g. provider_id://artists).
+        """
+        # Browse your provider's recommendations/media items.
+        # This is only called if you reported the BROWSE feature in the supported_features.
+        # You should return a list of MediaItems or ItemMappings for the given path.
+        # Note that you can return nested levels with BrowseFolder items.
+
+        # The MusicProvider base model has a default implementation of this method
+        # that will call the get_library_* methods if you did not override it.
+        return []
+
+    async def recommendations(self) -> list[RecommendationFolder]:
+        """
+        Get this provider's recommendations.
+
+        Returns an actual (and often personalised) list of recommendations
+        from this provider for the user/account.
+        """
+        # Get this provider's recommendations.
+        # This is only called if you reported the RECOMMENDATIONS feature in the supported_features.
+        return []
+
+    async def sync_library(self, media_type: MediaType) -> None:
+        """Run library sync for this provider."""
+        # Run a full sync of the library for the given media type.
+        # This is called by the music controller to sync items from your provider to the library.
+        # As a generic rule of thumb the default implementation within the MusicProvider
+        # base model should be sufficient for most (streaming) providers.
+        # If you need to do some custom sync logic, you can override this method.
+        # For example the filesystem provider in MA, overrides this method to scan the filesystem.
diff --git a/music_assistant/providers/refreezer/manifest.json b/music_assistant/providers/refreezer/manifest.json
new file mode 100644
index 00000000..6d5ace3f
--- /dev/null
+++ b/music_assistant/providers/refreezer/manifest.json
@@ -0,0 +1,10 @@
+{
+  "type": "music",
+  "domain": "refreezer",
+  "name": "ReFreezer",
+  "description": "An alternative Deezer music streaming & downloading client, based on Freezer",
+  "documentation": "https://github.com/DJDoubleD/refreezer",
+  "codeowners": ["custom"],
+  "requirements": [],
+  "multi_instance": true
+}
