diff --git a/music_assistant/providers/spotify/__init__.py b/music_assistant/providers/spotify/__init__.py
index a7e77e40..4f1e35d3 100644
--- a/music_assistant/providers/spotify/__init__.py
+++ b/music_assistant/providers/spotify/__init__.py
@@ -59,7 +59,8 @@ async def get_config_entries(
                 "code_challenge_method": "S256",
                 "code_challenge": code_challenge,
                 "redirect_uri": CALLBACK_REDIRECT_URL,
-                "state": auth_helper.callback_url,
+                # "state": auth_helper.callback_url,
+                "state": f"{values.get(CONF_CALLBACK_URL)}/callback/{auth_helper.session_id}",
             }
             query_string = urlencode(params)
             url = f"https://accounts.spotify.com/authorize?{query_string}"
@@ -109,10 +110,18 @@ async def get_config_entries(
             key=CONF_REFRESH_TOKEN,
             type=ConfigEntryType.SECURE_STRING,
             label=CONF_REFRESH_TOKEN,
-            hidden=True,
             required=True,
             value=values.get(CONF_REFRESH_TOKEN) if values else None,
         ),
+        ConfigEntry(
+            key=CONF_CALLBACK_URL,
+            type=ConfigEntryType.STRING,
+            label="callback url",
+            required=True,
+            default_value=mass.webserver.base_url.replace(':8095', ':8097', 1),
+            value=values.get(CONF_CALLBACK_URL) if values else mass.webserver.base_url.replace(':8095', ':8097', 1),
+            hidden=not auth_required,
+        ),
         ConfigEntry(
             key=CONF_CLIENT_ID,
             type=ConfigEntryType.SECURE_STRING,
